### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by t1nglanwang**

### CODE REVIEW

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

File: 2023/homework/TinglanWang/homework_matplotlib.ipynb
Comment (iphysresearch) [inline 2023/homework/TinglanWang/homework_matplotlib.ipynb:1]: 完成Matplotlib数据可视化作业！

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

File: 2023/homework/TinglanWang/homework_credit_scoring.ipynb
Comment (iphysresearch) [inline 2023/homework/TinglanWang/homework_credit_scoring.ipynb:1]: 完成seaborn建模作业

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 99
- Score: 91.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

File: 2023/homework/TinglanWang/homework_credit_scoring.ipynb
Comment (iphysresearch) [inline 2023/homework/TinglanWang/homework_credit_scoring.ipynb:1]: 完成模型搭建作业！

File: 2023/homework/TinglanWang/homework_credit_scoring_finetune_ensemble.ipynb
Comment (iphysresearch) [inline 2023/homework/TinglanWang/homework_credit_scoring_finetune_ensemble.ipynb:1]: 完成模型调优作业！

**Comment by iphysresearch** [comment]
@t1nglanwang 
话说Python编程的基础作业还有几个题有误哈~ 另外Python扩展作业也做一下哈？

**Comment by t1nglanwang** [comment]
> @t1nglanwang 话说Python编程的基础作业还有几个题有误哈~ 另外Python扩展作业也做一下哈？

哦哦，好的老师，谢谢提醒！ @iphysresearch

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 105
- Score: 97.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 108
- Score: 100.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 9
- Score: 90.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 108
- Score: 100.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 10
- Score: 100.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 11
- Score: 91.00%
---

**Comment by iphysresearch** [comment]
Good Job!!!   @t1nglanwang
---
Python homework:
- Total questions: 108
- Correct answers: 108
- Score: 100.00%
---
Numpy homework:
- Total questions: 10
- Correct answers: 10
- Score: 100.00%
---
Pandas homework:
- Total questions: 12
- Correct answers: 12
- Score: 100.00%
---

**Comment by t1nglanwang** [comment]
434. Number of Segments in a String
```class Solution:
    def countSegments(self, s: str) -> int:
        flag = 0 ##表示接下来会有新单词
        num = 0
        for i in range(len(s)):
            if flag==0 and s[i] != ' ':#如果前一个为空格，且该字符为字母，则单词个数加1，并改变flag
                flag = 1 # 表示接下来为空格
                num += 1
            if flag==1 and s[i] == ' ':
                flag = 0

        return num
```

**Comment by t1nglanwang** [comment]
1869. Longer Contiguous Segments of Ones than Zeros
```
class Solution:
    def checkZeroOnes(self, s: str) -> bool:
        maxlen1 = 0 #存储最大值
        maxlen0 = 0
        flaglen1 = 0#存储当前值
        flaglen0 = 0

        for i in s:
            if i == '1':#处理当前为1的情况
                flaglen1 += 1#如果当前为1则加1,直到不为1
            else:#不为1之后将当前长度与最长进行比较，取最大值。并将当前值归零。
                if flaglen1 > maxlen1:
                    maxlen1 = flaglen1

                flaglen1=0

            if i == '0':#同1
                flaglen0 += 1
            
            else:
                if flaglen0 > maxlen0:
                    maxlen0 = flaglen0

                flaglen0=0
        if flaglen1 > maxlen1:
                    maxlen1 = flaglen1
        if flaglen0 > maxlen0:
                    maxlen0 = flaglen0

        if maxlen1 > maxlen0:
            return True
        else:
            return False
        
```

**Comment by t1nglanwang** [comment]
1784. Check if Binary String Has at Most One Segment of Ones

```
class Solution:
    def checkOnesSegment(self, s: str) -> bool:
        flag = 1#记录是否切换状态的label
        num = 1#无前导0,故连续1的个数的初始值设为1
        for i in s:
            if flag == 0 and i=='1':#如果之前为0,且该字符为1,则加1
                num+=1

            if i=='0':如果为0,则flag设为0
                flag = 0
        
        if num >1:
            return False
        else:
            return True
```

**Comment by t1nglanwang** [comment]
852. Peak Index in a Mountain Array
```
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        #本质为找到最大值的index,所以先找出最大值，然后得到index。python默认的排序方法和二分查找时间复杂度均为O(log(n))，满足题意。
        max_value = sorted(arr,reverse=True)[0]
        index = arr.index(max_value)
        return index
```

**Comment by t1nglanwang** [comment]
162. Find Peak Element

```
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        if len(nums)==1:#过滤掉长度为1的，他们一定是peak
            return 0
        sorted_arr = sorted(nums,reverse = True)
        index_max = nums.index(sorted_arr[0])#最大值一定是一个peak

        return index_max
```

**Comment by t1nglanwang** [comment]
@iphysresearch 老师我的python部分的作业改错改完了，以及扩展作业也在评论区评论了，麻烦您有时间的时候看一下。谢谢！

**Review by iphysresearch — APPROVED** [review]
@t1nglanwang 
恭喜完成Python编程的基础作业和扩展作业，以及Matplotlib数据可视化和sklearn建模和调优作业！

**Pull request merged**